Implement "feed" command for CoolMasterNet as a service

Rjevski opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This PR implements a new service that allows Home Assistant to provide an ambient temperature value to CoolMasterNet-controlled HVAC units. This is useful in case your HVAC's built-in temperature sensor isn't optimally-located and is reporting inaccurate values and you want to replace it with a HA-connected one. You can set up an automation to react to your temperature sensor's state changes and call the new service.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.


- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
-->

This PR upgrades `pycoolmasternet-async` to 0.1.3 for which I've [opened a PR](https://github.com/OnFreund/pycoolmasternet-async/pull/3) in its repo. This has now been merged & the resulting release published to PyPi, so we are good to go.

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Rjevski commented on homeassistant/components/coolmaster/services.yaml:
```diff
@@ -0,0 +1,18 @@
+      description: The ambient temperature value to send to the unit.
+      selector:
+        number:
+          min: -70
+          max: 122
```
The idea is to accommodate both Celsius and Fahrenheit here as the interpretation of the value will be determined by how the CoolMasterNet device is configured.

Rjevski commented on homeassistant/components/coolmaster/services.yaml:
```diff
@@ -0,0 +1,18 @@
+    temperature:
+      name: Ambient temperature
+      description: The ambient temperature value to send to the unit.
+      selector:
+        number:
+          min: -70
+          max: 122
+          step: 0.1
+          mode: box
```
This is copied from the `climate.set_temperature` service.

OnFreund reviewed (approved):
LGTM

OnFreund commented:
@Rjevski Please add a documentation PR

Rjevski commented:
Documentation has been updated in https://github.com/home-assistant/home-assistant.io/pull/22145.

Rjevski commented:
Is there anything outstanding on this that I would need to do in order to get this merged in one of the next releases?

frenck commented:
@Rjevski 👋  I've noticed this PR is stuck and has conflicts now.

Could you please resolve the merge conflict in this PR?

Additionally, I noticed this PR bumps a dependency. However, could you please update the PR description to contain at least one (or multiple) of the following:

- A link to the release notes of this package version, and all versions in between.
- A link to the changelog of this package.
- A link to a Git(Hub) diff/compare view from the current version to the bumped version.

This allows us to review upstream changes and document them for historical purposes.

Thanks already! 👍

../Frenck

Rjevski commented:
Apologies for not getting back to you earlier - I've been quite busy with work and other life commitments.

Since opening the PR I've forked and rewritten the `pycoolmasternet-async` library and published it on [GitHub](https://github.com/rjevski/pycoolmasternet-ng) and on [PyPi](https://pypi.org/project/pycoolmasternet-ng/) as well.

In addition, I've updated the `coolmaster` component [in my branch](https://github.com/Rjevski/core/tree/dev/homeassistant/components/coolmaster) to use the new library. It introduces some new functionality:

* other transports - in addition to telnet, serial is now supported and the code is structured in such a way that other transports can be added down the line (REST and maybe even CoolAutomation's official cloud)
* reporting of AC heating/cooling demand if supported (faked if not - this can be removed or set behind a flag)
* AC unit capability (heat, cool, dry, etc) as well as display name is now configured at the CoolMasterNet gateway level using `set` commands - this means that the configuration persists with the gateway and is independent of HA. I also believe that's how CoolAutomation's cloud-based app does it, so it would make switching to Home Assistant easier for users of the official app (and in fact they would be able to coexist)
* reporting of AC unit error status
* reporting of filter status and a button to reset it
* support for the "feed" command to provide ambient temperature in the form of a service (what this PR was about)
* support for the "fan swing" functionality, if the device supports it

Would this be of interest? I could clean that up for inclusion into HA, though due to the breaking changes I'd prefer if this was a new integration (as we can't migrate the unit capability settings which used to be set at the HA level - while we can technically issue `set` commands to the gateway to replicate the HA settings I'd rather not do that). If not just let me know and I can update this PR.

OnFreund commented:
Hey @Rjevski 

Other transports and some of the features are definitely interesting. Not sure why you think that `set` is a breaking change, or why HA and the cloud app can't coexist - I use both.
I definitely wouldn't go for a new integration - no point in keeping two. If you really think this is breaking, we can always come up with a flag in the config entry options to turn the breaking behavior on or off.

Rjevski commented:
@OnFreund I believe it's a breaking change because:

* the current integration configures unit capabilities at the HA level and entity ID is set to the "line identifier" (such as `L5.001`)
* the new integration configures capabilities & entity ID using variables persisted on the gateway itself

A user who's never set those variables would after upgrade have their units report incorrect capabilities in HA and the entity IDs are likely to change. Similarly, a user who until now was able to use the integration despite misconfigured unit properties on the CoolMasterNet gateway would suddenly have their HA integration not work since it would now honor the (incorrect) settings of the gateway. Either way, upgrading is likely to require manual intervention to reconfigure the automations/etc to the new entity IDs and I would this consider it a breaking change.

> why HA and the cloud app can't coexist - I use both.

They can but this requires the HA-level unit capabilities to match the CoolMasterNet gateway settings otherwise the unit names and capabilities may differ between what's in HA and the official CoolAutomation app.

OnFreund commented:
You're talking about two things:
1. Capabilities. I don't see the HA setting and coolmaster as mutually exclusive. The Coolmaster settings are what the unit is capable of. The HA settings are what the user would like to see in the UI. The intersection of both is what the UI should display for each unit. You can think of the HA settings as sort of a whitelist - only what's on it should appear, but it should only appear if the unit has the capability. While the user might see fewer options than today, the options that will be hidden are ones that weren't available anyway.
I definitely wouldn't advocate changing Coolmaster capabilities based on what the user chooses to see in HA.
2. Change to the entity ID. Is there a compelling argument for this change? I don't see a real reason to do that, and even the coolmaster app shows the line identifiers. You might want to change the name based on what you're getting from Coolmaster, but I see no reason to change the ID. Even if you do change the entity ID, as long as your keep the unique id the same (which definitely has no reason to change), it will not be breaking.

It's very uncommon for HA to have more than one integration for the same piece of hardware, and I don't see a compelling need here. I'm happy to test the integration with My Coolmasternet + Fujitsu after your changes, and there's also a beta period with every HA release.

Rjevski commented:
Update:

@OnFreund  I have published my "improved" integration as a custom integration for the time being - feel free to [try it out](https://github.com/Rjevski/ha-coolmaster-ng). If it ends up being useful we can look into merging it into core in a separate PR and address the concerns above.

@frenck I would like to get this merged regardless of the above discussion - this is a small change that could be useful to others. As per your request:

> A link to the release notes of this package version, and all versions in between.

https://github.com/OnFreund/pycoolmasternet-async/releases/tag/v0.1.3

> A link to the changelog of this package.

Couldn't see an "official" changelog on PyPi, but the above link should cover that.

> A link to a Git(Hub) diff/compare view from the current version to the bumped version.

https://github.com/OnFreund/pycoolmasternet-async/compare/b6264bd55ff78e27961948773822901e3ba3a71a...f690b4cb63e305fa9a3efe39607c470e1a0526e7

(I will update the PR to resolve conflicts shortly)

Rjevski committed:
Implement "feed" command as a service in CoolMasterNet integration

OnFreund commented:
Thanks @Rjevski. I'm pretty sure the core team would prefer the Feed functionality to be implemented as a separate entity (a [Number entity](https://developers.home-assistant.io/docs/core/entity/number) probably makes the most sense (it even has a device class of type `temperature`. I would also suggest making that entity disabled by default, and only people with units that support it (according to the coolmaster docs it seems pretty rare) will enable it.

Rjevski commented:
@OnFreund the problem with the "feed" command is that it gives you zero feedback whether a unit has been "fed" - it will just reflect in the ambient temperature returned for that unit by the CoolMasterNet device for some time until the cache on the CM expires.

Therefore the "number" entity's current value could potentially be misleading as there is no guarantee the *current* value is the one that has been "fed" as opposed to just the ambient temperature.

I've instead opted for a write-only approach by using a service to not have to deal with this problem, but happy to hear your/others' thoughts.

OnFreund commented:
I think that the whole point of the number entity is to show the number that the user picked, not the one from the device, but I'd rather defer to the core team on this.

Rjevski commented:
Unless we cache it locally (and that raises its own issues, as it wouldn't account for edge-cases such as using the "feed" command directly from another machine or another HA), there's no way to retrieve the number that has been "fed" from the CM device - the best you can do is look at the ambient temperature (the "feed" command will lock it to the provided value for a while) but there's no indication whether that value is the "true" ambient or the one that has been "fed".

Rjevski commented:
@OnFreund

> I think that the whole point of the number entity is to show the number that the user picked, not the one from the device

In this case, when should this value be sent to the CoolMasterNet device? CM will cache the value for a while but as far as I know there are no further details. There is no way for us to know whether there is a currently set "feed" value in the CM device. We could send the value at regular intervals, but this could potentially interfere with other users of the CM device (if something else - think building automation controllers, etc) is also using it in addition to Home Assistant. You'd also need a way to clear the value which I don't believe Number entities support?

The idea of having it as a service is that it's a feature for advanced users that pretty much just passes through the request straight to CM - it will be up to the user to decide how to integrate it with their system (just avoids them having to create a shell command and using netcat/etc to send the value to the CM's IP address directly - that's how I've done it in the past). Adding layers of abstraction on top raises more questions/edge-cases and might interfere with users' own setups.

Thoughts?

OnFreund commented:
OK, I guess I don't understand the feed mechanism well enough.
